File: typescript/src/ai/openAI/clients/azureOpenAIClientAbstract.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { ILogger } from '../../../utils/logger';
import { IAzureDeployments } from '../httpSchema';
import { OpenAIClientAbstract } from './openAIClientAbstract';

export abstract class AzureOpenAIClientAbstract extends OpenAIClientAbstract {
    // Default Azure OpenAI REST API version
    protected static readonly DefaultAzureAPIVersion = '2022-12-01';

    // Azure endpoint of your models
    protected azureOpenAIApiVersion: string;

    // Azure endpoint of your models
    protected endpoint: string = '';

    constructor(log?: ILogger) {
        super(log);
        this.azureOpenAIApiVersion = AzureOpenAIClientAbstract.DefaultAzureAPIVersion;
    }

    protected async getDeploymentNameAsync(modelId: string): Promise<string | undefined> {
        const fullModelId = `${this.endpoint}:${modelId}`;

        // If the value is a deployment name
        if (s_deploymentToModel.has(fullModelId)) {
            return modelId;
        }

        // If the value is a model ID present in the cache
        if (s_modelToDeployment.has(fullModelId)) {
            return s_modelToDeployment.get(fullModelId);
        }

        // If the cache has already been warmed up
        let modelsAvailable: string;
        if (s_deploymentsCached.has(this.endpoint)) {
            modelsAvailable = Array.from(s_modelToDeployment.keys()).join(', ');
            throw new Error(
                `Model '${modelId}' not available on ${this.endpoint}. ` +
                    `Available models: ${modelsAvailable}. Deploy the model and restart the application.`
            );
        }

        await this.cacheDeploymentsAsync();

        if (s_modelToDeployment.has(fullModelId)) {
            return s_modelToDeployment.get(fullModelId);
        }

        modelsAvailable = Array.from(s_modelToDeployment.keys()).join(', ');
        throw new Error(
            `Model '${modelId}' not available on ${this.endpoint}. ` +
                `Available models: ${modelsAvailable}. Deploy the model and restart the application.`
        );
    }

    protected async cacheDeploymentsAsync(): Promise<void> {
        const url = `${this.endpoint}/openai/deployments?api-version=${this.azureOpenAIApiVersion}`;
        const response = await this.httpClient.get<IAzureDeployments>(url);
        if (response.status >= 400) {
            throw new Error(
                `Unable to fetch the list of model deployments from Azure. Status code: ${response.status}`
            );
        }

        if (!response.data || !Array.isArray(response.data.data) || response.data.data.length == 0) {
            throw new Error('Model not available. Unable to fetch the list of models.');
        }

        try {
            for (const deployment of response.data.data) {
                if (deployment.status != 'succeeded' || !deployment.model || !deployment.id) {
                    continue;
                }

                s_deploymentToModel.set(`${this.endpoint}:${deployment.id}`, deployment.model);
                s_modelToDeployment.set(`${this.endpoint}:${deployment.model}`, deployment.id);
            }
        } catch (err: any) {
            throw new Error(`Model not available. Unable to fetch the list of models: ${(err as Error).toString()}`);
        }

        s_deploymentsCached.set(this.endpoint, true);
    }
}

// Caching Azure details across multiple instances so we don't have to use "deployment names"
const s_deploymentsCached = new Map<string, boolean>();
const s_deploymentToModel = new Map<string, string>();
const s_modelToDeployment = new Map<string, string>();

File: dotnet/src/SemanticKernel/Connectors/OpenAI/AzureOpenAIClientAbstract.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Concurrent;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel.AI;
using Microsoft.SemanticKernel.Reliability;
using Microsoft.SemanticKernel.Text;

namespace Microsoft.SemanticKernel.Connectors.OpenAI;

/// <summary>
/// An abstract Azure OpenAI Client.
/// </summary>
public abstract class AzureOpenAIClientAbstract : OpenAIClientAbstract
{
    /// <summary>
    /// Default Azure OpenAI REST API version
    /// </summary>
    protected const string DefaultAzureAPIVersion = "2022-12-01";

    /// <summary>
    /// Azure OpenAI API version
    /// </summary>
    protected string AzureOpenAIApiVersion
    {
        get => this._azureOpenAIApiVersion;
        set
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                throw new AIException(
                    AIException.ErrorCodes.InvalidConfiguration,
                    "Invalid Azure OpenAI API version, the value is empty");
            }

            this._azureOpenAIApiVersion = value;
        }
    }

    /// <summary>
    /// Azure endpoint of your models
    /// </summary>
    protected string Endpoint { get; set; } = string.Empty;

    /// <summary>
    /// Construct an AzureOpenAIClientAbstract object
    /// </summary>
    /// <param name="log">Logger</param>
    /// <param name="handlerFactory">Retry handler factory</param>
    protected AzureOpenAIClientAbstract(ILogger? log = null, IDelegatingHandlerFactory? handlerFactory = null) : base(log, handlerFactory)
    {
    }

    /// <summary>
    /// Returns the deployment name of the model ID
    /// </summary>
    /// <param name="modelId">Azure OpenAI Model ID</param>
    /// <returns>Name of the deployment for the model ID</returns>
    /// <exception cref="AIException">AIException thrown during request.</exception>
    protected async Task<string> GetDeploymentNameAsync(string modelId)
    {
        string fullModelId = this.Endpoint + ":" + modelId;

        // If the value is a deployment name
        if (s_deploymentToModel.ContainsKey(fullModelId))
        {
            return modelId;
        }

        // If the value is a model ID present in the cache
        if (s_modelToDeployment.TryGetValue(fullModelId, out string modelIdCached))
        {
            return modelIdCached;
        }

        // If the cache has already been warmed up
        string modelsAvailable;
        if (s_deploymentsCached.ContainsKey(this.Endpoint))
        {
            modelsAvailable = string.Join(", ", s_modelToDeployment.Keys);
            throw new AIException(
                AIException.ErrorCodes.ModelNotAvailable,
                $"Model '{modelId}' not available on {this.Endpoint}. " +
                $"Available models: {modelsAvailable}. Deploy the model and restart the application.");
        }

        await this.CacheDeploymentsAsync();

        if (s_modelToDeployment.TryGetValue(fullModelId, out string modelIdAfterCache))
        {
            return modelIdAfterCache;
        }

        modelsAvailable = string.Join(", ", s_modelToDeployment.Keys);
        throw new AIException(
            AIException.ErrorCodes.ModelNotAvailable,
            $"Model '{modelId}' not available on {this.Endpoint}. " +
            $"Available models: {modelsAvailable}. Deploy the model and restart the application.");
    }

    /// <summary>
    /// Caches the list of deployments in Azure OpenAI.
    /// </summary>
    /// <returns>An async task</returns>
    /// <exception cref="AIException">AIException thrown during the request.</exception>
    protected async Task CacheDeploymentsAsync()
    {
        var url = $"{this.Endpoint}/openai/deployments?api-version={this.AzureOpenAIApiVersion}";
        HttpResponseMessage response = await this.HTTPClient.GetAsync(url);
        string json = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new AIException(
                AIException.ErrorCodes.ModelNotAvailable,
                $"Unable to fetch the list of model deployments from Azure. Status code: {response.StatusCode}",
                this.GetErrorMessageFromResponse(json));
        }

        lock (s_deploymentToModel)
        {
            try
            {
                var data = Json.Deserialize<AzureDeployments>(json);
                if (data == null)
                {
                    throw new AIException(
                        AIException.ErrorCodes.InvalidResponseContent,
                        "Model not available. Unable to fetch the list of models.");
                }

                foreach (var deployment in data.Deployments)
                {
                    if (!deployment.IsAvailableDeployment() || string.IsNullOrEmpty(deployment.ModelName) ||
                        string.IsNullOrEmpty(deployment.DeploymentName))
                    {
                        continue;
                    }

                    s_deploymentToModel[this.Endpoint + ":" + deployment.DeploymentName] = deployment.ModelName;
                    s_modelToDeployment[this.Endpoint + ":" + deployment.ModelName] = deployment.DeploymentName;
                }
            }
            catch (Exception e) when (e is not AIException)
            {
                throw new AIException(
                    AIException.ErrorCodes.UnknownError,
                    "Model not available. Unable to fetch the list of models.", e);
            }
        }

        s_deploymentsCached[this.Endpoint] = true;
    }

    #region private ================================================================================

    // Caching Azure details across multiple instances so we don't have to use "deployment names"
    private static readonly ConcurrentDictionary<string, bool> s_deploymentsCached = new();
    private static readonly ConcurrentDictionary<string, string> s_deploymentToModel = new();
    private static readonly ConcurrentDictionary<string, string> s_modelToDeployment = new();

    private string _azureOpenAIApiVersion = DefaultAzureAPIVersion;

    #endregion private ================================================================================
}
